### PR TITLE
Flag repository cfg client and flag directory

### PR DESCRIPTION
**Pull request by hakonhall**
- Makes new FlagRepository config server client to retrieve all flag data.
 - Makes WireFlagDataList to be used for creating the HTTP response in the
   config server, and parse the HTTP response in host admin.
 - Fixes problem with URL generation for controller: when port is not present
   in the request, remove ":-1" port specification in the url.
 - Makes a new FlagDirectory class, responsible for reading flags from
   /opt/vespa/var/vespa/flags (in FlagData JSON format), and reversely,
   sync that directory to exactly match a set of FlagData.
 - No longer have 'State state()' as a default method in interface.

### CODE REVIEW

File: flags/src/main/java/com/yahoo/vespa/flags/file/FlagDirectory.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+            return Files.list(flagDirectory).filter(Files::isRegularFile);
+        } catch (NotDirectoryException | NoSuchFileException e) {
+            return Stream.empty();
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private static String getFilenameOf(Path path) {
+        return path.getName(path.getNameCount() - 1).toString();
```
Comment (freva) [inline flags/src/main/java/com/yahoo/vespa/flags/file/FlagDirectory.java]: Why not `Path::getFileName`?

File: flags/src/main/java/com/yahoo/vespa/flags/file/FlagDirectory.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            uncheck(() -> Files.deleteIfExists(path));
+            modified = true;
+        }
+
+        return modified;
+    }
+
+    private Stream<Path> getAllRegularFilesStream() {
+        try {
+            return Files.list(flagDirectory).filter(Files::isRegularFile);
```
Comment (freva) [inline flags/src/main/java/com/yahoo/vespa/flags/file/FlagDirectory.java]: This stream needs to be closed or we will get too many open files. See VESPA-11886 and VESPA-12303

File: flags/src/main/java/com/yahoo/vespa/flags/file/FlagDirectory.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    public Optional<FlagData> readFlagData(FlagId flagId) {
+        return readUtf8File(getPathFor(flagId)).map(FlagData::deserialize);
+    }
+
+    /**
+     * Modify the flag directory as necessary, such that a later {@link #read()} will return {@code flagData}.
+     *
+     * @return true if any modifications were done.
+     */
+    public boolean sync(Map<FlagId, FlagData> flagData) {
```
Comment (freva) [inline flags/src/main/java/com/yahoo/vespa/flags/file/FlagDirectory.java]: Why not just save the entire thing in 1 file?

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/ConfigServerClients.java [hunk 4]
```diff
@@ -18,7 +19,10 @@
     Orchestrator orchestrator();
 
     /** Get handle to the /state/v1 REST API */
-    default State state() { throw new UnsupportedOperationException(); }
+    State state();
+
+    /** Get handle to the /flags/v1 REST API */
+    FlagRepository flagRepository();
```
Comment (freva) [inline node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/ConfigServerClients.java]: This needs to be implemented in `hosted` code

File: node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/ConfigServerClients.java [hunk 5]
```diff
@@ -18,7 +19,10 @@
     Orchestrator orchestrator();
 
     /** Get handle to the /state/v1 REST API */
-    default State state() { throw new UnsupportedOperationException(); }
+    State state();
```
Comment (freva) [inline node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/ConfigServerClients.java]: Needs to be implemented in `hosted` code

File: flags/src/main/java/com/yahoo/vespa/flags/file/FlagDirectory.java (hunk 1 above)
Comment (hakonhall) [reply flags/src/main/java/com/yahoo/vespa/flags/file/FlagDirectory.java]: omg, you have no idea how wide I searched for that - I don't know how I missed it.

**Review by freva — APPROVED** [review]

**Pull request merged**